### PR TITLE
fix(exo)!: extra undeclared args dropped

### PR DESCRIPTION
**Pull request by erights**
Alternative to https://github.com/endojs/endo/pull/1764 experiment.

Of the options listed there, this one

> Preserve the tolerant behavior, but not pass the values of the extra args through to the protected raw behavior method. This would preserve the input validation intent of method guards while still accommodating API growth and thus version slippage.

As suggested by @mhofman and @turadg

### CODE REVIEW

**Comment by erights** [comment]
Reviewers,

I made this a `fix(exo)!:` rather than a `fix(exo):` because it technically is a compat break. However, it is unlikely to currently be a compat break in practice, so let me know if you think I should remove the `!`.

File: packages/exo/src/exo-tools.js [hunk 1]
```diff
@@ -32,6 +32,13 @@ const { defineProperties, fromEntries } = Object;
  */
 const MinMethodGuard = M.call().rest(M.any()).returns(M.any());
 
+/**
+ * @param {Passable[]} args
+ * @param {MethodGuardPayload} methodGuardPayload
+ * @param {string | undefined} label
```
Comment (turadg) [inline packages/exo/src/exo-tools.js:38]: consider JSDoc convention:
```suggestion
 * @param {string} [label]
```

File: packages/exo/src/exo-tools.js [hunk 2]
```diff
@@ -40,6 +47,15 @@ const defendSyncArgs = (args, methodGuardPayload, label) => {
...
   mustMatch(harden(args), paramsPattern, label);
+  if (restArgGuard !== undefined) {
+    return args;
+  }
+  const declaredLen =
+    argGuards.length + (optionalArgGuards ? optionalArgGuards.length : 0);
+  if (args.length <= declaredLen) {
+    return args;
+  }
+  return args.slice(0, declaredLen);
```
Comment (turadg) [inline packages/exo/src/exo-tools.js:58]: consider documenting,
```suggestion
  // ignore extraneous arguments, as a JS function call would do
  return args.slice(0, declaredLen);
```

File: packages/exo/src/exo-tools.js [hunk 3]
```diff
@@ -53,8 +69,8 @@ const defendSyncMethod = (method, methodGuardPayload, label) => {
   const { syncMethod } = {
     // Note purposeful use of `this` and concise method syntax
     syncMethod(...args) {
-      defendSyncArgs(harden(args), methodGuardPayload, label);
-      const result = apply(method, this, args);
+      const realArgs = defendSyncArgs(harden(args), methodGuardPayload, label);
```
Comment (turadg) [inline packages/exo/src/exo-tools.js:72]: please also update the signature to qualify the args
```suggestion
      const realArgs = defendSyncArgs(harden(rawArgs), methodGuardPayload, label);
```

File: packages/exo/test/test-heap-classes.js [hunk 4]
```diff
@@ -10,6 +10,19 @@ import {
...
+  foo: M.call().returns(),
+});
+
+test('what happens with extra arguments', t => {
+  const exo = makeExo('WithExtra', NoExtraI, {
+    foo(x) {
+      t.is(x, undefined);
+    },
+  });
+  exo.foo(8);
```
Comment (turadg) [inline packages/exo/test/test-heap-classes.js:23]: takes a bit to know this value is irrelevant.
```suggestion
  exo.foo('an extra arg');
```

**Review by turadg — COMMENTED** [review]
I agree that a) this is technically a breaking change and b) it's unlikely to break anything.

Since the package version hasn't reached 1.0, there's little downside to making a breaking change. So I think keep the `!` to err on the side of being loud about it.

File: packages/exo/src/exo-tools.js (hunk 1 above)
Comment (erights) [reply packages/exo/src/exo-tools.js]: I had not intended to make it optional, but rather, merely allow its value to be `undefined`. It isn't exported and its two callers in this file provide the `label` argument explicitly.

However, I don't mind making it optional, so done.

File: packages/exo/src/exo-tools.js (hunk 2 above)
Comment (erights) [reply packages/exo/src/exo-tools.js:59]: Done.

File: packages/exo/src/exo-tools.js (hunk 3 above)
Comment (erights) [reply packages/exo/src/exo-tools.js:73]: I don't understand. There is no `rawArgs` in scope here. Do you suggest that I rename the syncMethod rest parameter to `rawArgs`?

> please also update the signature to qualify the args

I don't understand this either. The returned `syncMethod` as a whole is declared as `Method`. What do you have in mind?

File: packages/exo/test/test-heap-classes.js (hunk 4 above)
Comment (erights) [reply packages/exo/test/test-heap-classes.js]: Done

File: packages/exo/src/exo-tools.js (hunk 3 above)
Comment (mhofman) [reply packages/exo/src/exo-tools.js:73]: I believe the request is to rename `args` -> `rawArgs` to make it clearer what args represent

**Review by mhofman — APPROVED** [review]
LGTM

File: packages/exo/src/exo-tools.js (hunk 3 above)
Comment (turadg) [reply packages/exo/src/exo-tools.js:73]: right. I would have edited both lines but the "suggestion" tool only works on lines modified in the PR.
```js
    syncMethod(...rawArgs) {
      const realArgs = defendSyncArgs(harden(rawArgs), methodGuardPayload, label);
```

Comment (erights) [reply packages/exo/src/exo-tools.js]: Gotcha. This rename makes a different distinction than I had in mind, but a better one. Accordingly, I renamed this one and the existing occurrences of `rawArgs`, and one more, to `syncArgs`. Reads much better. Done.

**Pull request merged**